sigal/plugins/nomedia.py

```diff
@@ -0,0 +1,91 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+
+""" This plugin offers more fine-grained control over exluded images and folders, similarly to how
+    it's handled on Android.
+
+    To ignore a folder or image put a .nomedia file next to it in its parent folder and put its name
+    into the file.
+    E.g.:
+        content of folder:
+            IMG_3425.JPG, IMG_2426.JPG, IMG_2427.JPG, subfolder, .nomedia
+        content of .nomedia:
+            IMG_2426.JPG
+            IMG_2427.JPG
+            subfolder
+    will ignore all images but IMG_3425.JPG and the subfolder
+
+    Alternatively, if you put a .nomedia file into a folder and leave it blank (i.e. an empty file
+    called .nomedia in a folder containing images), this ignores the whole folder it's located in
+    (like on Android).
+
+    WARNING: When you have a pre-existing gallery from a previous run of sigal adding a new .nomedia
+    file will not remove the newly ignored images/albums from the existing gallery (only the entries
+    in the parent gallery pointing to it). They might still be reachable thereafter. Either remove
+    the whole gallery to be sure or remove the ignored files/folders inside the gallery to remove
+    them for good.
+"""
+
+import logging
+import os
+from sigal import signals
+
+logger = logging.getLogger(__name__)
+
+def filter_nomedia(album, settings=None):
+    """Removes all filtered Media and subdirs from an Album"""
+    nomediaPath = album.src_path + "/.nomedia"
+
+    if os.path.isfile(nomediaPath):
+
+        if os.path.getsize(nomediaPath) == 0:
+            logger.info("Ignoring album '{}' because of present 0-byte .nomedia file".\
+                            format(album.name))
+
+            #subdirs have been added to the gallery already, remove them there, too
+            for key in list(album.gallery.albums.keys()):
+                if key.startswith(album.path):
+                    del album.gallery.albums[key]
+
+            #cannot set albums => empty subdirs so that no albums are generated
+            album.subdirs = []
+            album.medias = []
+
+        else:
+            with open(nomediaPath, "Ur") as nomediaFile:
+                logger.info("Found a .nomedia file in {}, ignoring its entries".format(album.name))
+                ignoredEntries = nomediaFile.read().split("\n")
+
+                album.medias = list(filter(lambda media: media.filename not in ignoredEntries,
+                                            album.medias))
+                album.subdirs = list(filter(lambda dirName: dirName not in ignoredEntries,
+                                            album.subdirs))
+
+                #subdirs have been added to the gallery already, remove them there, too
+                for ignoredEntry in ignoredEntries:
+                    for key in list(album.gallery.albums.keys()):
```

Comment:
Without the explicit list-call you'd change the dict while iterating and that throws an exception.